src/spinnaker.cpp

```diff
@@ -0,0 +1,1060 @@
+        }
+        if (IsWritable(camera_->BinningVertical)) {
+            set_int_node(camera_->BinningVertical, (int64_t)target);
+        }
+        Spinnaker::GenApi::IInteger& binning = get_binning_node(camera_);
+        last_known_settings_.binning = (uint8_t)binning();
+    }
+}
+
+void
+SpinnakerCamera::maybe_set_sample_type(SampleType target)
+{
+    CHECK(target < SampleTypeCount);
+    if (target != last_known_settings_.pixel_type) {
+        set_enum_node(camera_->PixelFormat, to_pixel_format(target));
+        last_known_settings_.pixel_type =
+          to_sample_type(*(camera_->PixelFormat));
+    }
+}
+
+void
+SpinnakerCamera::maybe_set_offset(
+  CameraProperties::camera_properties_offset_s target)
+{
+    CameraProperties::camera_properties_offset_s& last =
+      last_known_settings_.offset;
+    if (target.x != last.x) {
+        set_int_node(camera_->OffsetX, (int64_t)target.x);
+        last.x = (uint32_t)camera_->OffsetX();
+    }
+    if (target.y != last.y) {
+        set_int_node(camera_->OffsetY, (int64_t)target.y);
+        last.y = (uint32_t)camera_->OffsetY();
+    }
+}
+
+void
+SpinnakerCamera::maybe_set_shape(
+  CameraProperties::camera_properties_shape_s target)
+{
+    CameraProperties::camera_properties_shape_s& last =
+      last_known_settings_.shape;
+    if (target.x != last.x) {
+        set_int_node(camera_->Width, (int64_t)target.x);
+        last.x = (uint32_t)camera_->Width();
+    }
+    if (target.y != last.y) {
+        set_int_node(camera_->Height, (int64_t)target.y);
+        last.y = (uint32_t)camera_->Height();
+    }
+}
+
+void
+SpinnakerCamera::update_input_trigger(Trigger& trigger)
+{
+    trigger.kind = Signal_Input;
+    trigger.enable = *(camera_->TriggerMode) == genicam_on;
+    trigger.line = to_trigger_line(*(camera_->TriggerSource));
+    trigger.edge = to_trigger_edge(*(camera_->TriggerActivation));
+}
+
+void
+SpinnakerCamera::maybe_set_input_trigger_frame_start(Trigger& target)
+{
+    if (!is_equal(target, last_known_settings_.input_triggers.frame_start)) {
+        // Always disable trigger before any other configuration as in the
+        // Spinnaker Trigger.cpp example.
+        set_enum_node(camera_->TriggerMode, genicam_off);
+
+        set_enum_node(camera_->TriggerSelector, genicam_frame_start);
+        set_enum_node(camera_->TriggerSource, to_trigger_source(target.line));
+        set_enum_node(camera_->TriggerActivation,
+                      to_trigger_activation(target.edge));
+        set_enum_node(camera_->TriggerMode,
+                      target.enable ? genicam_on : genicam_off);
+
+        update_input_trigger(last_known_settings_.input_triggers.frame_start);
+    }
+}
+
+void
+SpinnakerCamera::update_output_trigger_exposure(Trigger& trigger)
+{
+    trigger.kind = Signal_Output;
+    trigger.enable = (*(camera_->LineSource) == genicam_exposure_active) &&
+                     (*(camera_->LineSource) == genicam_exposure_active);
```

Comment:
Why is this here twice?

Reply:
Great question. I don't know. I swear at some point I was comparing two different things here.

Given the current hardcoding of triggers for the blackfly camera in this PR, I think this should probably be checking that the LineSelector is line 1, so I made that change. This will likely change quite a bit after adding support for the oryx camera.